### PR TITLE
Added SASL/OAuthBearer

### PR DESCRIPTION
**Pull request by oulydna**
### Changes

Fixes issue #618

Added `OAUTHBEARER` as a new `sasl_mechanism`. The implementation is mainly based on [this PR in kafka-python](https://github.com/dpkp/kafka-python/pull/1750).

### Checklist

- [x] I think the code is well written
- [x] Documentation reflects the changes
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

**Comment by oulydna** [comment]
Thanks @tvoinarovskyi. Really appreciate the quick reply. 

Yes, I tested it in our own environment. 

Good point on the async callback. I will look into this.

File: aiokafka/conn.py [hunk 1]
```diff
@@ -692,3 +717,47 @@ def create_salted_password(self, salt, iterations):
...
+
+        resp = yield data, True
+
+        assert resp == b"", (
+            "Server should either close or send an empty response"
+        )
+
+    def _build_oauth_client_request(self):
+        return "n,,\x01auth=Bearer {}{}\x01\x01".format(
+            self._sasl_oauth_token_provider.token(), self._token_extensions()
```
[the remark this answers is not in the document]
Comment (oulydna) [reply aiokafka/conn.py]: shouldn't we return None based on self._token_sent?

**Comment by oulydna** [comment]
@tvoinarovskyi can you please have a look at the new changes?

File: aiokafka/abc.py [hunk 2]
```diff
@@ -87,6 +90,17 @@ def on_partitions_assigned(self, assigned):
         pass
 
 
+class AbstractTokenProvider(BaseAbstractTokenProvider):
+    async def token(self):
```
Comment (tvoinarovskyi) [inline aiokafka/abc.py:94]: I think only providing a token async function with a proper specification is enough, let's remove the `run_in_executor`. By introducing the `_token` abstract function below you essentially require the developer to override it (you can't initialize an `abc` class without overriding all abstract methods), which is strange in the case where he only needs an async callback.
We can add a docstring describing how you can use synchronous callback instead.

File: aiokafka/abc.py [hunk 3]
```diff
@@ -87,6 +90,17 @@ def on_partitions_assigned(self, assigned):
...
+        return await asyncio.get_running_loop().run_in_executor(
+            None, self._token)
+
+    @abc.abstractmethod
+    def _token(self):
+        pass
+
```
Comment (tvoinarovskyi) [inline aiokafka/abc.py:101]: Please also add an `extensions` method, but leave it returning an empty sequence (not abstract method). This way the user would be able to understand how he can add extensions, but will not be required to do so.

**Comment by tvoinarovskyi** [comment]
@oulydna Looks good for a merge, few minor points if you have time and let's ship it 🚢

**Comment by tvoinarovskyi** [comment]
Thanks again!

**Comment by oulydna** [comment]
Thanks you @tvoinarovskyi . Really glad to have this integrated. We like aiokafka, and will like it better.

**Pull request merged**